### PR TITLE
Load OpenSSL 3 legacy provider explicitly in digestmd5 plugin

### PR DESCRIPTION
**Pull request by simo5**
Add support to load legacy provider directly in digestmd5 plugin so that rc4 is available.
This depends on #653 to be pushed first as it includes the same commits.
Addresses some of the points in #665

### CODE REVIEW

**Comment by simo5** [comment]
Marking as WIP because I do not have a test that exercises this plugin, to assure it works as desired

**Comment by simo5** [comment]
This seems to be working fine in Fedora Rawhide

**Comment by Neustradamus** [comment]
@quanah, @hyc: What do you think?
A merging before the 2.1.28?

**Comment by simo5** [comment]
Incorporated changes on #653 and therefore also rebased on master

**Review by hyc — CHANGES_REQUESTED** [review]
relying on OpenSSL's RC4 seems to be going the wrong direction, since it's optional there and frequently not built. Doing this properly requires adding configure tests for OpenSSL's RC4 support. That seems like a bunch of unnecessary hassle. Unless you know of actual flaws in the RC4 impl here, I'd drop this commit.

**Comment by simo5** [comment]
The reason to do this is that we actually want RC4 to be disable-able if needed in the distribution, and that is easier to do in one place (openssl) than trying to chase many different components.
I can change the code if you think we can pursue, but if you really want to drop it let me know, and I'll just maintain a downstream patch for it for now.

**Comment by quanah** [comment]
Need to see if there are any mechanisms for 2.2 that will use RC4.  If not, this PR can be closed.

File: plugins/digestmd5.c [hunk 1]
```diff
@@ -1102,60 +1114,139 @@ static int init_des(context_t *text,
...
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+    ossl3_context_t *ossl3_ctx;
 
-    rc = EVP_EncryptInit_ex(ctx, EVP_rc4(), NULL, enckey, NULL);
-    if (rc != 1) return SASL_FAIL;
+    rc = init_ossl3_ctx(text);
+    if (rc != SASL_OK) return rc;
+
+    ossl3_ctx = (ossl3_context_t *)text->crypto.libctx;
+    cipher = EVP_CIPHER_fetch(ossl3_ctx->libctx, "RC4", "");
```
Comment (mistotebe) [inline plugins/digestmd5.c:1211]: `cipher` is leaked if we got it from here as seen in openldap's test076-authid-rewrite (just build with `-fsanitize=address` or run the first `ldapwhoami` under valgrind).

Comment (simo5) [reply plugins/digestmd5.c:1211]: nice catch

**Comment by simo5** [comment]
@quanah should we just close this PR ?

File: plugins/digestmd5.c (hunk 1 above)
Comment (simo5) [reply plugins/digestmd5.c:1211]: rebased and fixed

Comment (mistotebe) [reply plugins/digestmd5.c:1211]: Thanks, @bgermann could you update `0029-Load-OpenSSL3-legacy-provider-digestmd5.patch` in Debian accordingly?

Comment (bgermann) [reply plugins/digestmd5.c:1211]: I won't patch an upstream patch. Please remind me to add this patch once it has arrived in master branch.

**Pull request closed without merging**